tower-route/src/lib.rs

```diff
@@ -0,0 +1,43 @@
+extern crate futures;
+extern crate tower_layer;
+extern crate tower_service;
+extern crate void;
+
+use futures::{future::Either, Async, Poll};
+use tower_service::Service;
+
+pub trait Acceptor<Request> {
+    fn accepts(&self, request: &Request) -> bool;
+}
+
+pub struct Route<A, B> {
+    service: A,
+    other: B,
+}
+
+impl<A, B, Request, Future, Response, Error> Service<Request> for Route<A, B>
+where
+    Future: futures::Future<Item = Response, Error = Error>,
+    A: Service<Request, Future = Future, Response = Response, Error = Error> + Acceptor<Request>,
+    B: Service<Request, Future = Future, Response = Response, Error = Error>,
+{
+    type Response = Response;
+    type Error = Error;
+    type Future = Either<A::Future, B::Future>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        match self.service.poll_ready() {
```

Comment:
For readiness, the strategy to use is poll_ready is always ready, and the response future uses Oneshot in tower-util.

Reply:
IIUC, why would you want to want to wrap the services with Oneshot? I'll have to borrow the services out of their `Route` struct to construct it.

Reply:
The problem with `Oneshot` is that it must own the services inside the route in order to construct it. So the services must also implement `Copy`. Is that ok?